src/storage/kv/transaction.rs

```diff
@@ -459,6 +466,73 @@ impl Transaction {
         self.read_set.lock().clear();
         self.snapshot.take();
     }
+
+    /// After calling this method the subsequent modifications within this
+    /// transaction can be rolled back by calling [`rollback_to_savepoint`].
+    ///
+    /// This method is stackable and can be called multiple times with the
+    /// corresponding calls to [`rollback_to_savepoint`].
+    ///
+    /// [`rollback_to_savepoint`]: Transaction::rollback_to_savepoint
+    pub fn set_savepoint(&mut self) -> Result<()> {
+        // If the transaction mode is not mutable (i.e., it's read-only), return an error.
+        if !self.mode.mutable() {
+            return Err(Error::TransactionReadOnly);
+        }
+        // If the transaction is closed, return an error.
+        if self.closed {
+            return Err(Error::TransactionClosed);
+        }
+
+        // Bump the latest savepoint number.
+        self.savepoints += 1;
+
+        Ok(())
+    }
+
+    /// Rollback the state of the transaction to the latest savepoint set by
+    /// calling [`set_savepoint`].
+    ///
+    /// [`set_savepoint`]: Transaction::set_savepoint
+    pub fn rollback_to_savepoint(&mut self) -> Result<()> {
+        // If the transaction mode is not mutable (i.e., it's read-only), return an error.
+        if !self.mode.mutable() {
+            return Err(Error::TransactionReadOnly);
+        }
+        // If the transaction is closed, return an error.
+        if self.closed {
+            return Err(Error::TransactionClosed);
+        }
+        // Check that the savepoint is set
+        if self.savepoints == 0 {
+            return Err(Error::TransactionWithoutSavepoint);
+        }
+
+        // Remove entries marked for rollback since the last call
+        // to set_savepoint() from the write set, the write order map
+        // and the snapshot.
+        self.write_set.retain(|e| {
```

Comment:
Just thinking of the scenario where a scan was done on a range, and then rollbacked, I think we would need to track the read_key_ranges with savepoints too

Reply:
Would the read_set not required to be rollbacked too? Say I read a few keys since the last savepoint, now those keys in the read_set are used in conflict_detection. Wouldn't rolling back to a particular savepoint imply removing the tracked keys post the savepoint?

Reply:
Yes, you're right, I think we should track and clean up read set too on rollback.